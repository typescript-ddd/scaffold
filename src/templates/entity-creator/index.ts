export * from "./entity-creator.generator";
export * from "./entity-creator.template";
export * from "./entity-creator.types";
