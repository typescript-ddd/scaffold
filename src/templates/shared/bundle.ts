import { GenerateContext, Chunk } from "./generate-context";
import { Template, TemplateOption } from "./template";

export interface Bundle<
  TValues extends { [K in keyof TOptions]: any },
  TOptions extends { [key: string]: TemplateOption }
> extends Omit<Template<TValues, TOptions>, "generate"> {
  generate(value: TValues, context: GenerateContext): Chunk[];
}
