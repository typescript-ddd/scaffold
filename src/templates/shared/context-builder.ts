import { GenerateContext, ContextImportDeclaration } from "./generate-context";
import { arrays } from "../../utils";
import { SourceFile } from "ts-morph";

export class ContextBuilder implements GenerateContext {
  private _fileName: string;
  private _importDeclarations: ContextImportDeclaration[] = [];

  private constructor(
    private rootDir: string = "@/src",
    fileName: string = "example.ts"
  ) {
    this._fileName = fileName;
  }

  resolveDir(
    type: "domain" | "infrastructure" | "application" | "presentation",
    subPath?: string
  ): string {
    switch (type) {
      case "domain":
        return `${this.rootDir}/domain${subPath ? `/${subPath}` : ""}`;
      case "infrastructure":
        return `${this.rootDir}/infrastructure${subPath ? `/${subPath}` : ""}`;
      case "application":
        return `${this.rootDir}/application${subPath ? `/${subPath}` : ""}`;
      case "presentation":
        return `${this.rootDir}/presentation${subPath ? `/${subPath}` : ""}`;
      default:
        throw new Error(`Unknown type ${type} for resolveDir`);
    }
  }

  getImportDeclarations(moduleSpecifier?: string): ContextImportDeclaration[] {
    const modules = this.getModuleSpecifiers();

    const declarations: ContextImportDeclaration[] = [];

    for (const currentModule of modules) {
      const namedImports = this.getNamedImports(currentModule);
      declarations.push({ moduleSpecifier: currentModule, namedImports });
    }

    if (moduleSpecifier) {
      return declarations.filter((x) => x.moduleSpecifier === moduleSpecifier);
    }

    return declarations;
  }

  addImportDeclaration(declaration: ContextImportDeclaration): void {
    this._importDeclarations.push(declaration);
  }

  getModuleSpecifiers(): string[] {
    return arrays.distinct(
      this._importDeclarations.map((x) => x.moduleSpecifier)
    );
  }

  getNamedImports(moduleSpecifier: string): string[] {
    return arrays.distinct(
      this._importDeclarations
        .filter((x) => x.moduleSpecifier === moduleSpecifier)
        .map((x) => x.namedImports)
        .flat()
    );
  }

  writeSourceFileImports(sourceFile: SourceFile) {
    const importDeclarations = this.getImportDeclarations();
    const moduleSpecifiers = this.getModuleSpecifiers();

    const coreImports = importDeclarations.filter(
      (x) => x.moduleSpecifier === "@typescript-ddd/core"
    );

    if (coreImports) {
      sourceFile.addImportDeclaration({
        moduleSpecifier: `@typescript-ddd/core`,
        namedImports: coreImports.map((x) => x.namedImports).flat(),
      });
    }

    // Non core and root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter(
        (x) => x !== "@typescript-ddd/core" && !x.startsWith(this.rootDir)
      )
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }

    // Root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter((x) => x.startsWith(this.rootDir))
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }
  }

  clearImportDeclarations(): void {
    this._importDeclarations = [];
  }

  tokenize(text: string, tokens: { name: string }[]): string {
    let output = text;
    tokens.forEach((token) => {
      output = output.replace(`{{${token.name}}}`, "${" + token.name + "}");
    });
    return "`" + output + "`";
  }

  get fileName(): string {
    return this._fileName;
  }

  static build(rootDir: string): ContextBuilder {
    return new ContextBuilder(rootDir);
  }
}
