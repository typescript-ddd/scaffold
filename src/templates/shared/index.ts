export * from "./context-builder";
export * from "./generate-context";