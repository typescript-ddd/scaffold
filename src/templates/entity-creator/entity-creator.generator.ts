import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { Chunk, GenerateContext } from "../shared";
import { EntityCreatorTemplateValues } from "./entity-creator.types";

export const generateEntityCreator = (
  values: EntityCreatorTemplateValues,
  context: GenerateContext,
  chunkName?: string
): Chunk => {
  const { entityName } = values;
  const entityCreatorType = `${strings.capitalize(entityName)}Creator`;
  const repositoryType = `${strings.capitalize(entityName)}Repository`;
  const entityType = strings.capitalize(entityName);
  const entityIdType = `${entityType}Id`;
  const createPropsType = `Create${entityType}Props`;
  const {
    fileName = `${strings.kebab(entityName, "creator")}.ts`,
    projectPath = ["application", "create"],
  } = context.currentFile || {};
  const project = new Project();
  const sourceFile = project.createSourceFile(
    fileName,
    `
export class ${entityCreatorType} {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["DomainEventBus", "Uuid"],
  });

  context.addImportDeclaration({
    moduleSpecifier: context.resolveDir("domain"),
    namedImports: [entityType, entityIdType, createPropsType, repositoryType],
  });

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(entityCreatorType);
  classDeclaration.addJsDoc({
    description: `Represents ${strings.describe(entityCreatorType)}.`,
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    scope: Scope.Private,
    parameters: [
      {
        name: "repository",
        type: repositoryType,
        scope: Scope.Private,
        isReadonly: true,
      },
      {
        name: "domainEventBus",
        type: "DomainEventBus",
        scope: Scope.Private,
        isReadonly: true,
      },
    ],
  });
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      entityCreatorType
    )}.`,
    tags: [
      {
        tagName: "param",
        text: `{${repositoryType}} repository - The ${strings.natural(
          entityName
        )} repository.`,
      },
      {
        tagName: "param",
        text: `{DomainEventBus} domainEventBus - The domain event bus.`,
      },
    ],
  });

  classDeclaration
    .addMethod({
      name: "create",
      scope: Scope.Public,
      returnType: `Promise<${entityType}>`,
      parameters: [
        {
          name: "props",
          type: createPropsType,
        },
      ],
      isAsync: true,
    })
    .setBodyText((writer) => {
      writer.writeLine(
        `const ${strings.lower(
          entityName
        )} = ${entityType}.create(props, ${entityIdType}.create());`
      );
      writer.writeLine(
        `await this.repository.save(${strings.lower(entityName)});`
      );
      writer.writeLine(
        `const events = ${strings.lower(entityName)}.getUncommittedEvents();`
      );
      writer.writeLine(`this.domainEventBus.publish(...events);`);
      writer.writeLine(`return ${strings.lower(entityName)};`);
    })
    .addJsDoc({
      description: `Creates a new ${strings.natural(entityName)}.`,
      tags: [
        {
          tagName: "param",
          text: `{${createPropsType}} props - The properties to create the ${strings.natural(
            entityName
          )}.`,
        },
        {
          tagName: "returns",
          text: `{Promise<${entityType}>} The created ${strings.natural(
            entityName
          )}.`,
        },
      ],
    });

  return {
    name: chunkName || "EntityCreator",
    fileName,
    projectPath,
    content: sourceFile.getFullText(),
    values,
  };
};
