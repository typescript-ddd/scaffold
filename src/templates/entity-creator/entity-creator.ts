import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";

export type GenerateEntityCreatorOptions = {
  entityName: string;
};

export const generateEntityCreator = (
  options: GenerateEntityCreatorOptions,
  context: GenerateContext
) => {
  const { entityName } = options;
  const entityCreatorType = `${strings.capitalize(entityName)}Creator`;
  const repositoryType = `${strings.capitalize(entityName)}Repository`;
  const entityType = strings.capitalize(entityName);
  const entityIdType = `${entityType}Id`;
  const createPropsType = `Create${entityType}Props`;

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${context.fileName}.ts`,
    `
export class ${entityCreatorType} {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["DomainEventBus", "Uuid"],
  });

  context.addImportDeclaration({
    moduleSpecifier: context.resolveDir("domain"),
    namedImports: [entityType, entityIdType, createPropsType, repositoryType],
  });

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(entityCreatorType);
  classDeclaration.addJsDoc({
    description: `Represents ${strings.describe(entityCreatorType)}.`,
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    scope: Scope.Private,
    parameters: [
      {
        name: "repository",
        type: repositoryType,
        scope: Scope.Private,
        isReadonly: true,
      },
      {
        name: "domainEventBus",
        type: "DomainEventBus",
        scope: Scope.Private,
        isReadonly: true,
      },
    ],
  });
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      entityCreatorType
    )}.`,
    tags: [
      {
        tagName: "param",
        text: `{${repositoryType}} repository - The ${strings.natural(
          entityName
        )} repository.`,
      },
      {
        tagName: "param",
        text: `{DomainEventBus} domainEventBus - The domain event bus.`,
      }
    ],
  });

  classDeclaration.addMethod({
    name: "create",
    scope: Scope.Public,
    returnType: `Promise<${entityType}>`,
    parameters: [
      {
        name: "props",
        type: createPropsType,
      },
    ],
    isAsync: true,
  }).setBodyText((writer) => {
    writer.writeLine(`const ${strings.lower(entityName)} = ${entityType}.create(props, ${entityIdType}.create());`);
    writer.writeLine(`await this.repository.save(${strings.lower(entityName)});`);
    writer.writeLine(`const events = ${strings.lower(entityName)}.getUncommittedEvents();`);
    writer.writeLine(`this.domainEventBus.publish(...events);`);
    writer.writeLine(`return ${strings.lower(entityName)};`);
  }).addJsDoc({
    description: `Creates a new ${strings.natural(entityName)}.`,
    tags: [
      {
        tagName: "param",
        text: `{${createPropsType}} props - The properties to create the ${strings.natural(entityName)}.`,
      },
      {
        tagName: "returns",
        text: `{Promise<${entityType}>} The created ${strings.natural(entityName)}.`,
      },
    ],
  })

  let output = sourceFile.getFullText();

  return output;
};
