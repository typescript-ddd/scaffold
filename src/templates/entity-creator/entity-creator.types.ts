import { TemplateStringOption } from "../shared/template";

export type EntityCreatorTemplateValues = {
  entityName: string;
};

export type EntityCreatorTemplateOptions = {
  entityName: TemplateStringOption;
};
