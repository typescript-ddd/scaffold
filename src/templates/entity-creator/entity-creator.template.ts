import { GenerateContext } from "../shared";
import { Template } from "../shared/template";
import { generateEntityCreator } from "./entity-creator.generator";
import {
  EntityCreatorTemplateOptions,
  EntityCreatorTemplateValues,
} from "./entity-creator.types";

export class EntityCreatorTemplate
  implements
    Template<EntityCreatorTemplateValues, EntityCreatorTemplateOptions>
{
  readonly name = "Entity Creator";
  readonly description = "Generates an entity creator";
  readonly options: EntityCreatorTemplateOptions = {
    entityName: {
      name: "entityName",
      label: "Entity Name",
      description: "The name of the entity",
      optionType: "string",
    },
  };
  readonly defaultValues = {
    entityName: "User",
  };
  generate(values: EntityCreatorTemplateValues, context: GenerateContext): string {
    return generateEntityCreator(values, context);
  }
}
