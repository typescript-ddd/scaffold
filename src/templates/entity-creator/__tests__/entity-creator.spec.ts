import { GenerateContext, ContextBuilder } from "../../shared";
import { generateEntityCreator } from "../entity-creator";

describe("EntityCreator", () => {
  let context: GenerateContext;

  beforeEach(() => {
    context = ContextBuilder.build("@src");
  });

  it("Should render the entity creator template", () => {
    const output = generateEntityCreator(
      {
        entityName: "User",
      },
      context
    );

    expect(output).toMatchSnapshot();
  });
});
