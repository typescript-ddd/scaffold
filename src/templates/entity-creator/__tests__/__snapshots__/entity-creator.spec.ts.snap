// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntityCreator Should render the entity creator template 1`] = `
"import { DomainEventBus, Uuid } from "@typescript-ddd/core";
import { User, UserId, CreateUserProps, UserRepository } from "@src/domain";

/** Represents a user creator. */
export class UserCreator {
    /**
     * Initializes a new instance of a user creator.
     * @param {UserRepository} repository - The user repository.
     * @param {DomainEventBus} domainEventBus - The domain event bus.
     */
    private constructor(private readonly repository: UserRepository, private readonly domainEventBus: DomainEventBus) {
    }

    /**
     * Creates a new user.
     * @param {CreateUserProps} props - The properties to create the user.
     * @returns {Promise<User>} The created user.
     */
    public async create(props: CreateUserProps): Promise<User> {
        const user = User.create(props, UserId.create());
        await this.repository.save(user);
        const events = user.getUncommittedEvents();
        this.domainEventBus.publish(...events);
        return user;
    }
}    
"
`;
